server/apps/immich/src/utils/disk-storage.ts

```diff
@@ -0,0 +1,101 @@
+import { Request } from 'express';
+import { ParamsDictionary } from 'express-serve-static-core';
+import { DiskStorageOptions, StorageEngine } from 'multer';
+import * as crypto from 'node:crypto';
+import * as fs from 'node:fs';
+import * as os from 'node:os';
+import * as path from 'node:path';
+import { PassThrough } from 'node:stream';
+import type { ParsedQs } from 'qs';
+
+type GetFileNameFn = (req: Request, file: Express.Multer.File, cb: (error: Error | null, filename: string) => void) => void;
+type GetDestFn = (req: Request, file: Express.Multer.File, cb: (error: Error | null, destination: string) => void) => void;
+
+export interface ExtendedDiskStorageOptions extends DiskStorageOptions {
+}
+
+function getFilename(
+  req: Request,
+  file: Express.Multer.File,
+  cb: (error: Error | null, filename: string) => void
+) {
+  crypto.randomBytes(16, function (err, raw) {
+    cb(err, err ? '' : raw.toString('hex'));
+  });
+}
+
+function getDestination(
+  req: Request,
+  file: Express.Multer.File,
+  cb: (error: Error | null, destination: string) => void
+) {
+  cb(null, os.tmpdir())
+}
+
+class ExtendedDiskStorage implements StorageEngine {
+  private _getFilenameFn: GetFileNameFn;
+  private _getDestinationFn: GetDestFn;
+
+  constructor(opts: ExtendedDiskStorageOptions) {
+    this._getFilenameFn = (opts.filename || getFilename);
+
+    if (typeof opts.destination === 'string') {
+      fs.mkdirSync(opts.destination, { recursive: true });
+      this._getDestinationFn = ($0, $1, cb) => {
+        cb(null, opts.destination as string);
+      };
+    } else {
+      this._getDestinationFn = (opts.destination || getDestination);
+    }
+  }
+
+  public _handleFile(req: Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>, file: Express.Multer.File, cb: (error?: any, info?: Partial<Express.Multer.File> | undefined) => void): void {
+    const that = this;
```

Comment:
Why this? or "that"?

Reply:
To make sure callback use `this` pointer right way. I just keep multer implement for stable.

Reply:
So anything that pointed from `that` is your extended implementation of Multer, correct?

Reply:
Not everything, as you can see at line 62 in [server/apps/immich/src/utils/disk-storage.ts](https://github.com/immich-app/immich/pull/525/files/02bf73f1f9eb0a7a07ab05a38ade6d3b6a3a180e#diff-0479d6afcc07a79a3f2c54eaae31177bfab0ef9776d1d54bcd4d478f4e6824d3) is my implement. I try keep code base from multer. I also really don't like callback implementation that Multer used. Can we change in async method?

Reply:
I think let's try to stick with Multer's way of doing things for now, so we don't interfere too much with the library.